ext/uri/php_uri.c

```diff
@@ -311,6 +436,99 @@ static void uri_equals(INTERNAL_FUNCTION_PARAMETERS, zend_object *that_object, z
 	zend_string_release(that_str);
 }
 
+PHP_METHOD(Uri_Rfc3986_Uri, equals)
+{
+	zend_object *that_object;
+	zend_object *comparison_mode = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(1, 2)
+		Z_PARAM_OBJ_OF_CLASS(that_object, uri_rfc3986_uri_ce)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS(comparison_mode, uri_comparison_mode_ce);
+	ZEND_PARSE_PARAMETERS_END();
+
+	uri_equals(INTERNAL_FUNCTION_PARAM_PASSTHRU, that_object, comparison_mode);
+}
+
+PHP_METHOD(Uri_Rfc3986_Uri, toRawString)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	zend_object *this_object = Z_OBJ_P(ZEND_THIS);
+	uri_internal_t *internal_uri = uri_internal_from_obj(this_object);
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	zend_string *uri_str = internal_uri->handler->uri_to_string(internal_uri->uri, URI_RECOMPOSITION_RAW_ASCII, false);
+	if (uri_str == NULL) {
+		zend_throw_exception_ex(NULL, 0, "Cannot recompose %s to string", ZSTR_VAL(this_object->ce->name));
+		RETURN_THROWS();
+	}
+
+	RETURN_STR(uri_str);
+}
+
+PHP_METHOD(Uri_Rfc3986_Uri, toString)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	zend_object *this_object = Z_OBJ_P(ZEND_THIS);
+	uri_internal_t *internal_uri = uri_internal_from_obj(this_object);
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	zend_string *uri_str = internal_uri->handler->uri_to_string(internal_uri->uri, URI_RECOMPOSITION_NORMALIZED_ASCII, false);
+	if (uri_str == NULL) {
+		zend_throw_exception_ex(NULL, 0, "Cannot recompose %s to string", ZSTR_VAL(this_object->ce->name));
+		RETURN_THROWS();
+	}
+
+	RETURN_STR(uri_str);
+}
+
+PHP_METHOD(Uri_Rfc3986_Uri, resolve)
+{
+	zend_string *uri_str;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1) \
```

Comment:
Backslashes left over from macros probably?

Reply:
Yes, most probably that was the case